### PR TITLE
Implement simple example using swash to render to png

### PR DESCRIPTION
**Pull request by nicoburns**
This is basically working. But it could definitely do with some actual review and cleanup before merging.

## Output

output

## Notes

- ~Emoji rendering is not working~
- Font fallback for arabic text does seem to be working (note: I pulled the arabic text from cosmic text's sample files. Google translate translate the included arabic text as "arabic", but I cannot otherwise vouch for it's content).
- Subpixel rendering is not enabled (does this make sense when rendering to image?)
- The rendered text doesn't look fantastic. Possibly my blending code (used to apply the Swash alpha mask to an RGBA ImageBuffer from the image buffer is incorrect / too naive?). Cosmic-text pulls in tiny-skia in it's examples. I was hoping to avoid that here, but perhaps it's necessary?
- ~Converting from a `parley::Font` to a `swash::FontRef` is very awkward, specifically when the font file is a collection. The Parley `Font` stores the index of the font within the collection, but Swash's `FontRef` expect the byte offset of the font within the collection. I could not find this exposed anywhere in parley/fontique/skrifa/read-fonts so I'm currently doing pointer arithmetic to get it.~
- ~I'm currently writing to a file called `output.png` in the users current working directory. Possibly we could do better than that (some kind of gitignored "data" directory in the parley repo?)~
- ~I've added an impl of `Brush` for `peniko::Color`, but possibly we ought to just blanket impl for all types implementing the requisite super traits?~
- ~I'm not using the `NormalizedCoords` as they always seemed to be empty in my example (what are they for?)~

### CODE REVIEW

File: examples/simple.rs
Comment (waywardmonkeys) [inline examples/simple.rs:74]: I think you can do something like:

```
let mut img = RgbaImage::from_pixel(width + (padding * 2), height + (padding * 2), Rgba([background_color, ....]));
```

File: examples/simple.rs
Comment (dfrg) [inline examples/simple.rs:120]: Swash has this method: https://docs.rs/swash/0.1.15/swash/struct.FontRef.html#method.from_index

**Comment by dfrg** [comment]
> Emoji rendering is not working

We lost this in the switch to fontique. I can submit a small patch to fix it. 

> Subpixel rendering is not enabled (does this make sense when rendering to image?)

Probably not unless we’re just doing it for effect. Might be nice to add an option at some point. 

> The rendered text doesn't look fantastic.

Black on white without gamma correction is going to look “dirty”. Something simple like `alpha.pow(1.0/2.2)` would improve this a bit. 

> I'm not using the `NormalizedCoords` as they always seemed to be empty in my example (what are they for?)

These are the parameters for controlling interpolation in variable fonts. You should be able to feed the slice directly into `ScalerBuilder::normalized_coords`.

**Comment by nicoburns** [comment]
> We lost this in the switch to fontique. I can submit a small patch to fix it.

Ah. Yes, it seems important!

> Black on white without gamma correction is going to look “dirty”. Something simple like `alpha.pow(1.0/2.2)` would improve this a bit.

Perhaps using these functions? https://observablehq.com/@toja/color-blending-with-gamma-correction.

If I'm understanding correctly the correct process is:

- Convert to float
- Convert to linear rgb
- Perform computation
- Convert srgb
- Convert to u8

> These are the parameters for controlling interpolation in variable fonts. You should be able to feed the slice directly into `ScalerBuilder::normalized_coords`.

I see. Does this also need to be plugged in with skrifa?

**Comment by dfrg** [comment]
See #56 for a patch that might get emoji working with this example

**Comment by dfrg** [comment]
> Perhaps using these functions? https://observablehq.com/@toja/color-blending-with-gamma-correction.
> 
> If I'm understanding correctly the correct process is:
> 
> * Convert to float
> * Convert to linear rgb
> * Perform computation
> * Convert srgb
> * Convert to u8

We can do full sRGB if you'd like but I think just applying the power function to the alpha component will give us reasonably good results for the example.

> > These are the parameters for controlling interpolation in variable fonts. You should be able to feed the slice directly into `ScalerBuilder::normalized_coords`.
> 
> I see. Does this also need to be plugged in with skrifa?

It does. We use the `LocationRef` type there which is really just a wrapper for `&[NormalizedCoord]`. This unfortunately requires a type conversion and allocation as is done in xilem/masonry. We should probably change parley to use this type in the future.

**Comment by nicoburns** [comment]
> We can do full sRGB if you'd like but I think just applying the power function to the alpha component will give us reasonably good results for the example.

The trouble I was having with apply `pow` to the alpha channel is that it's a u8 and I'm unable to apply a fractional power to a u8.

I have implemented an attempt at sRGB <-> Linear RGB conversion. Not sure if I've done it correctly as the result isn't obviously better (but it's good enough that it looks plausibly correct to me). I've left both functions in the code for the time being if you want to inspect them.

Naive:

output_naive

Gamma corrected:

output_gamma

> It does. We use the `LocationRef` type there which is really just a wrapper for `&[NormalizedCoord]`. This unfortunately requires a type conversion and allocation as is done in xilem/masonry. We should probably change parley to use this type in the future.

Ah, gotcha. I'll have a look at plugging this in tomorrow.

**Comment by dfrg** [comment]
Assuming this is using system-ui on macOS, setting the normalized coordinates should also get you the requested bold on the first four characters since San Francisco is a variable font.

**Comment by nicoburns** [comment]
> Assuming this is using system-ui on macOS, setting the normalized coordinates should also get you the requested bold on the first four characters since San Francisco is a variable font.

Ah, that would be good. Bold was working with Helvetica, but it generally looked pretty bad with Helvetica so I switched it out. And I'd like to leave it on `system-ui` as that will presumably pick something sensible cross-platform. I guess maybe I could also use `sans-serif`?

**Comment by nicoburns** [comment]
Pretty sure my gamma correction function is wrong as the serif font looks much better with the naive blending.

File: examples/simple.rs
Comment (dfrg) [inline examples/simple.rs:214]: Should be `glyph_width * 4` since we want the full stride for each row. The image will also need to be blended otherwise it’ll have a black background.

**Comment by dfrg** [comment]
> Pretty sure my gamma correction function is wrong as the serif font looks much better with the naive blending.

Yeah, this is exceptionally tricky to get right. I’ll play with it later this evening.

File: examples/simple.rs
Comment (nicoburns) [reply examples/simple.rs]: Ah yes. With this change (+the same blending as I'm using for the alpha mask + #56 applied) emojis are rendering reasonably.

output

**Review by waywardmonkeys — DISMISSED** [review]
I'm assuming that once the other (tiny-skia) lands, this will move to a crate as well?

**Comment by nicoburns** [comment]
> I'm assuming that once the other (tiny-skia) lands, this will move to a crate as well?

Yes. Hoping to land the tiny-skia one (#55) first. Will update this one once we've settled on a design there.

**Comment by nicoburns** [comment]
> > Pretty sure my gamma correction function is wrong as the serif font looks much better with the naive blending.
> 
> Yeah, this is exceptionally tricky to get right. I’ll play with it later this evening.

Hmm... the `image` crate (that we're already using here) has a blend function (https://docs.rs/image/latest/image/struct.Rgba.html#method.blend). Perhaps we could just use that. Looks the same as my "naive" version to me, but it certainly makes the example shorter.

**Comment by nicoburns** [comment]
> These are the parameters for controlling interpolation in variable fonts. You should be able to feed the slice directly into `ScalerBuilder::normalized_coords`.

@dfrg Hmm.. I've added this, but I still don't seem to get bold text.

**Comment by dfrg** [comment]
> > These are the parameters for controlling interpolation in variable fonts. You should be able to feed the slice directly into `ScalerBuilder::normalized_coords`.
> 
> @dfrg Hmm.. I've added this, but I still don't seem to get bold text.

Was just working on this :) Found a bug in swash's avar table. Update to swash 0.1.16 should fix it.

**Comment by dfrg** [comment]
Overall, this is looking pretty good to me now. I'm happy to punt on the gamma issue, get the tiny-skia version landed and then merge this one after making the crate extraction changes.

File: examples/simple.rs
Comment (nicoburns) [reply examples/simple.rs:73]: done

File: examples/swash_render/src/main.rs
Comment (nicoburns) [inline examples/swash_render/src/main.rs:165]: @dfrg What are the semantics of the `SubpixelMask`? (/ how would one apply that to an RGBA image?). I reckon it would be good to demonstrate that even if we don't actually enable subpixel mode in the example.

Comment (dfrg) [reply examples/swash_render/src/main.rs:165]: This is an rgba image with an alpha component per color channel (the alpha in the mask is ignored). This document (https://github.com/servo/webrender/blob/main/webrender/doc/text-rendering.md#subpixel-text-blending) has a good explanation of the math.

File: examples/swash_render/src/main.rs
Comment (nicoburns) [inline examples/swash_render/src/main.rs:183]: @dfrg Would you be open to adding this pixel iteration code into swash itself(I'm thinking as a method on `Mask` or `swash::Image`)? It would be a nice simplification for this example and could also be used in cosmic-text which has some similar iteration code.

Comment (nicoburns) [reply examples/swash_render/src/main.rs:183]: (I don't think we should block this PR on that)

Comment (dfrg) [reply examples/swash_render/src/main.rs:183]: Yes, that seems like a useful addition.

**Comment by nicoburns** [comment]
I've rebased this on top of latest `main` and made it fit the new directory structure. It is ready for another round of review.

File: examples/simple.rs
Comment (dfrg) [inline examples/simple.rs:92]: Should this match the output path of the tiny-skia example?

File: examples/simple.rs
Comment (dfrg) [inline examples/simple.rs:200]: Nit: the scaler settings are dependent on glyph run properties but invariant per glyph so we could build it once per run and pass it in here. This is what I would suggest in real code as a performance optimization so the example might want to reflect that.

File: examples/swash_render/src/main.rs
Comment (dfrg) [inline examples/swash_render/src/main.rs:1]: Is this duplicated intentionally?

**Review by dfrg — APPROVED** [review]
LGTM with a couple suggestions and some confusion about the code duplication.

File: examples/swash_render/src/main.rs
Comment (nicoburns) [reply examples/swash_render/src/main.rs:1]: No. This file is the correct one. The other file is a rebase conflict.

File: examples/simple.rs
Comment (nicoburns) [reply examples/simple.rs]: Fixed

**Pull request merged**
